Feature: add user management

Kore-rep opened the pull request:
Resolves https://github.com/councilwatch/councilwatch/issues/13
and https://github.com/councilwatch/councilwatch/issues/25

Commit:
Docs and infra

Commit:
RBAC

Commit:
Linting updates

Commit:
Remove global application of auth guards

Commit:
Users controller v1

Commit:
Add council Ids and various fixes

KasimAhmic commented on packages/server/src/users/users.controller.ts:
```diff
@@ -1,7 +1,76 @@
+  })
+  deleteSelf(@CurrentUser() user: User): Promise<User> {
+    return this.usersService.deleteUser(user.email);
+  }
```
No need for a dedicated endpoint for users to delete themselves. They'll have access to their internal ID and should just be able to pass that to the endpoint.

Kore-rep replied:
Agree, updated

KasimAhmic commented on packages/server/src/users/users.controller.ts:
```diff
@@ -1,7 +1,76 @@
+  })
+  deleteUser(@Query('email') userEmail: string): Promise<User> {
+    return this.usersService.deleteUser(userEmail);
+  }
```
We should follow REST patterns here and pass the ID in the path like:

```suggestion
  @Delete('/{id}')
  @ApiSecurity(JWT_SECURITY)
  @ApiOperation({
    summary: 'Delete user',
    description: 'Deletes a single user by the given ID.',
  })
  deleteUser(@Param('id') id: string, @CurrentUser() user: User): Promise<User> {
    return this.usersService.deleteUser(id, user);
  }
```

Additionally, the `UsersService` should handle the role check. If the user is a regular user, they can only delete themselves (`id === user.id`) but if they're a mod, they can delete whoever.

Kore-rep replied:
Updated

KasimAhmic commented on packages/server/src/users/users.controller.ts:
```diff
@@ -1,7 +1,76 @@
+  async getUserNotificationCriteria(@CurrentUser() user: User): Promise<string[]> {
+    const usr = await this.usersService.getUser(user.email, true);
+    return usr.councils;
+  }
```
Just return the entire User entity instead of returning a portion of it. We'll need the entire user for the UI anyway, two birds with one stone.

```suggestion
  @Get()
  @ApiSecurity(JWT_SECURITY)
  @ApiOperation({
    summary: 'Get a user',
    description: 'Returns the current user that is logged in.',
  })
  async getUser(@CurrentUser() user: User): Promise<string[]> {
    return this.usersService.getUser(user.email, true);
  }
```

KasimAhmic commented on packages/server/src/users/users.controller.ts:
```diff
@@ -1,7 +1,76 @@
+    @Body() criteria: UpdateCriteriaDto,
+  ): Promise<User> {
+    return this.usersService.updateUserNotificationCriteria(user, criteria);
+  }
```
This should just be a general update request where they can send an `UpdateUserDto` object to change any of the fields in the User entity. Then the service does a role check to see if they're authorized to change their role (i.e. if the user is a moderator). This way we also get rid of the `promoteUserToModerator` route.

KasimAhmic commented on biome.jsonc:
```diff
@@ -23,6 +23,12 @@
       "style": {
         "useImportType": "off" // https://biomejs.dev/linter/rules/use-import-type/#caveat-with-typescript-experimental-decorators
+      },
+      "correctness": {
+        "noUnusedImports": {
+          "level": "warn",
+          "fix": "safe"
+        }
```
Is this telling Biome to automatically delete unused imports?

Kore-rep replied:
Correct, can revert if needed

KasimAhmic commented on packages/server/src/users/users.module.ts:
```diff
@@ -1,13 +1,17 @@
 
 @Module({
-  imports: [TypeOrmModule.forFeature([User])],
+  imports: [TypeOrmModule.forFeature([User]), forwardRef(() => AuthModule), AppConfigModule, PassportModule],
```
Why do we need to import the AuthModule in the UsersModule?

Kore-rep replied:
Leftover from me figuring out stuff about Nest, fixed that now

KasimAhmic reviewed (commented):
The code itself looks great, but we can simplify a lot of this down to four endpoints:

```
GET /        Retrieves the current logged in user
GET /{id}    Retrieves a user by id (mod only)
PUT /{id}    Updates a user by id (only certain fields if user, most fields if mod)
DELETE /{id} Deletes a user by id (can only delete self if user, any user if mod)
```

Commit:
Simplify controller and refactor service

Commit:
Import removal

Kore-rep commented:
Refactored the service and cleaned up the controller. Made some improvements to role checking including blocking moderators from modifying admins and other moderators

KasimAhmic reviewed (approved):
LGTM!

Commit:
Merge branch 'main' into feature/user-management
